ansi: Designate G0 graphic character set

richiejp opened the pull request:
Hi, I am new to Rust so this may not be the best way to go about doing this, but I am trying to implement the line drawing characters for applications like YasT and GNU/Linux's `make menuconfig`. So far I have spent a lot of time reading specifications (only then to realise you have a nice neat summary in this repo) and have made some stubs. I haven't looked at remapping the glyphs yet. Also, great job!

Related issue: https://github.com/jwilm/alacritty/issues/226
Commit message:
Parse `ESC ( B` as designate the G0 graphic character set as ASCII and parse
`ESC ( 0` as designate the G0 graphic character set as line drawing
glyphs. This does not implement actually using the line drawing glyphs,
but prints out a debug message when the control sequence is parsed.

jwilm commented:
Thanks for the PR!

I've actually got something a bit different in mind for this feature.

The ANSI terminal protocol specifies a total of four charsets, G0-G3. The protocol may specify some character set to use for subsequent output, this is referred to as locking shift. It's also possible the terminal will specify some character set to be operated on for the following sequence (this is the part you've started to deal with) and then be followed with a specific character set ('B' refers to the graphics/line drawing mode).

I think the following methods would make sense for the `ansi::Handler`:

```rust
// Switch character set for output
fn set_charset(&mut self, CharsetIndex);

// Configure a character set
fn designate_charset(&mut self, CharsetIndex, Charset);
```

This would allow the terminal to record some state that, for example, the G0 character set is `DEC Special Character and Line Drawing Set`. Then in the input method, the handler can inspect this and remap characters as needed.

I'm referencing http://invisible-island.net/xterm/ctlseqs/ctlseqs.html btw.

How does that sound? Do you fancy continuing to work on this?

richiejp commented:
OK, sounds reasonable, I was ignoring the locking shift stuff because I haven't actually observed it being used and was a bit confused as to what the point of it is. If I have understood correctly then it might make sense (at least on 1980s hardware) for application writers to set G1 to line drawing then shift in (SI) when drawing lines and shift out (SO) to go back to text. Alternatively they could use Gn with LSn where _n_ is the set index or something even more complicated with multi-byte sets. Instead, so far at least, I have just seen apps set G0 to the line drawing set then back to ASCII/latin again. Hopefully what I am saying makes sense?

Anyway, I would like to continue working on this. I have limited time though, so if someone is desperate for it and I haven't been heard from just give me a kick.

jwilm commented:
> OK, sounds reasonable, I was ignoring the locking shift stuff because I haven't actually observed it being used and was a bit confused as to what the point of it is. If I have understood correctly then it might make sense (at least on 1980s hardware) for application writers to set G1 to line drawing then shift in (SI) when drawing lines and shift out (SO) to go back to text. Alternatively they could use Gn with LSn where n is the set index or something even more complicated with multi-byte sets. Instead, so far at least, I have just seen apps set G0 to the line drawing set then back to ASCII/latin again. Hopefully what I am saying makes sense?

Yeah that all makes sense, and I don't have any idea how often LS is used. It is pretty easy to support, though. The easiest thing is for the `Term` type to have an array of 4 charsets and the active charset:

```rust
/// Newtype for charsets array so we can prevent bad indexing statically
struct Charsets([Charset; 4]);

/// How charsets newtype is indexed
impl Index<CharsetIndex> for Charsets {
    type Output = Charset;
    fn index(&self, index: CharsetIndex) -> &Charset {
        &self.0[index as usize]
    }
}

/// Term description
struct Term {
    charset: [Charset; 4],
    active_charset: CharsetIndex,
    // .. everything else on Term
}
```

where `CharsetIndex` is probably an enum with variants like `G0, G1, ..`. Locking shift would just modify `active_charset` in this scenario.

The `ansi::Handler::input()` implementation for `Term` can then just lookup the active charset using those properties.

> Anyway, I would like to continue working on this. I have limited time though, so if someone is desperate for it and I haven't been heard from just give me a kick.

You are of course welcome and encouraged to do so! Please be aware though that this is pretty high on the list of features Alacritty needs, and such a kick is likely to come relatively soon.

richiejp commented:
So that will cover shifting char sets into the `GL` area. The `GR` area wouldn't be too difficult to support as well, but it overlaps with UTF-8 and probably isn't used much so I will ignore it for now.

richiejp commented:
What I have done so far works, but probably doesn't make the best use of Rust's type system. I will see if I can clear it up a bit and also add some tests.

jwilm commented on src/ansi.rs:
```diff
@@ -436,6 +442,23 @@ pub enum Attr {
+#[derive(Clone, Copy, Debug)]
+pub enum StandardCharset {
+    Ascii,
+    LineDrawing,
```
It turns out to include more than just line drawing glyphs. The long name is usually "Special character and line drawing". I'm not a huge fan of long names in most cases, but it feels appropriate here (there are other line drawing + special character charsets).

jwilm commented on src/ansi.rs:
```diff
@@ -742,13 +765,22 @@ impl<'a, H, W> vte::Perform for Performer<'a, H, W>
+        };
+
         match byte {
+            b'B' if i1 == b'(' =>
```
I was a bit worried about how the compiler would treat the match statement after this. It looks like we still get a jump table for each variant, and the cases with `if expr` just have a comparison added in. Nifty.

jwilm commented on src/term/mod.rs:
```diff
@@ -200,6 +200,65 @@ pub use self::mode::TermMode;
 pub const TAB_SPACES: usize = 8;
 
+#[derive(Clone, Copy)]
+enum Charset {
```
I'd like to hold off on this particular addition until there's a way to use it. For the time being, it would just be added complexity.

Instead, let's use the `enum Charset` defined in _ansi.rs_. For mapping inputs, we could add a trait impl for it in this module.

jwilm commented on src/term/mod.rs:
```diff
@@ -200,6 +200,65 @@ pub use self::mode::TermMode;
+            Charset::Ascii => c,
+            Charset::Sparse(set) => match set {
+                StandardCharset::LineDrawing =>
+                    match c {
```
We should use a lookup table for the line drawing set. I'm guessing we can find one in another terminal as long as the license is amenable.

richiejp replied:
I am not entirely sure what you mean, but from looking at the source for a few terminal emulators I found links to http://vt100.net/docs/vt102-ug/table5-13.html and http://www.mirrorservice.org/sites/www.bitsavers.org/pdf/dec/terminal/vt340/EK-VT3XX-TP-001_VT330_VT340_Text_Programming_Mar87.pdf (page 41 for technical character set). 

In terms of the implementation I was debating whether to use an array and index into it using the first byte of _c_ with an offset, or just use _c_ in a match statement, the latter definitely makes more sense if only a few characters are different from ASCII (sparse mapping), but maybe not if all the chars are different (dense mapping) or it needs to be configurable.

jwilm replied:
You can't safely use the first byte of `c`. Easiest thing to do is cast to `usize` and check that it's in the 7-bit ascii range we care about. Casting to a `usize` sets us up to index into an array (with an offset as you suggested). Although.. since we need to do some conditional logic anyway, maybe it makes sense for this character set to just use a `match`.

I would kind of like to test a couple of implementations since this will be such a hot code path. Unfortunately we're not set up to benchmark it at the moment.

Maybe for starters, just do whatever is easiest.

richiejp replied:
Also https://github.com/sourcelair/xterm.js/blob/master/src/xterm.js#L5001licensed under MIT

jwilm replied:
Here's a version I put together using the [table](http://vt100.net/docs/vt102-ug/table5-13.html) you provided. As such, it won't need attribution. I skipped some symbols that I was having trouble tracking down (HT, FF, CR, LF). You should double check that it's correct.

It's not totally comprehensive, but I think it covers the important stuff.

```rust
        match c as usize {
            '<' => '◆',
            '=' => '▒',
            'j' => '┘',
            'k' => '┐',
            'l' => '┌',
            'm' => '└',
            'n' => '┼',
            'o' => '⎺',
            'p' => '⎻',
            'q' => '─',
            'r' => '⎼',
            's' => '⎽',
            't' => '├',
            'u' => '┤',
            'v' => '┴',
            'w' => '┬',
            'x' => '│',
            'y' => '≤',
            'z' => '≥',
            '{' => 'π',
            '|' => '≠',
            '}' => '£',
            '~' => '·',
            _ => (), // TODO
        }
```

richiejp replied:
Yeah, I would be interested to see how the two approaches compare (at some point). If the Rust compiler/LLVM is clever enough they should be the same or the match statement should be faster.

jwilm replied:
The big unknown imo is how branch prediction ends up working for a typical workload. Once we're set up to benchmark, we can come back and play around with a table-based implementation.

jwilm commented on src/ansi.rs:
```diff
@@ -742,13 +765,22 @@ impl<'a, H, W> vte::Perform for Performer<'a, H, W>
             b'Z' => self.handler.identify_terminal(self.writer),
             b'c' => self.handler.reset_state(),
+            b'0' if i1 == b'(' =>
+                self.handler.designate_charset(CharsetIndex::G0, StandardCharset::LineDrawing),
```
Since these are getting wrapped, can we add `{` and `}`?

richiejp commented:
Thanks for the feedback! I have changed some stuff in the mean time, but it is all useful to me.

jwilm commented on src/term/mod.rs:
```diff
@@ -200,63 +200,74 @@ pub use self::mode::TermMode;
-            _ => c
+            StandardCharset::Ascii => c,
+            StandardCharset::SpecialCharacterAndLineDrawing =>
+                // XXX Should write a macro for defining character mappings
```
Wouldn't the macro end up looking basically like the `match` statement here?

jwilm commented on src/term/mod.rs:
```diff
@@ -200,63 +200,74 @@ pub use self::mode::TermMode;
+impl CharsetMapping for StandardCharset {
     /// Switch/Map character to the active charset. Ascii is the common case and
     /// for that we want to do as little as possible.
     #[inline]
```
This is a good idea, but in practice this won't be inlinable... unless we make the `fn input()` generic over `CharsetMapping`.

jwilm commented on src/term/mod.rs:
```diff
@@ -200,63 +200,74 @@ pub use self::mode::TermMode;
-    /// A user defined charset handled by some more complicated mechanism
-    User,
+trait CharsetMapping {
+    fn map(&self, c: char) -> char;
```
This could probably have a default impl that just returns `c`.

jwilm commented:
Btw, if you want quicker feedback on stuff, you are welcome to join our IRC channel: #alacritty on freenode.

richiejp commented:
I have squashed the commits, added a test and cleaned it up a bit. I'm not sure what to do about the inlining. I think it is probably best not to add more to it until someone requests it. I have tried a handful of apps with it and they seem to work fine, apart from some glitches which are related to other bugs.

I will join IRC if I get around to it :-) 👍

jwilm commented on src/term/mod.rs:
```diff
@@ -228,6 +301,9 @@ pub struct Term {
     /// Alt cursor
     alt_cursor: Point,
 
+    charsets: Charsets,
```
Would be nice to have some comments on these

jwilm commented on src/term/mod.rs:
```diff
@@ -1109,18 +1187,31 @@ impl ansi::Handler for Term {
     }
+
+    #[inline]
+    fn designate_charset(&mut self, index: CharsetIndex, charset: StandardCharset) {
```
Thought about the name a bit, and something like `configure_charset` seems much more intuitive. What do you think?

richiejp replied:
On the one hand I think 'configure' will instantly make more sense to most programmers, on the other 'designate' is more accurate and sticks to the specification. If we were rigorously implementing the entire specification I would go with designate, but otherwise changing it to 'configure' and documenting it seems more helpful.

jwilm commented on src/term/mod.rs:
```diff
@@ -1109,18 +1187,31 @@ impl ansi::Handler for Term {
+    }
+
+    #[inline]
+    fn set_charset(&mut self, index: CharsetIndex) {
```
Maybe `set_active_charset` would be slightly less ambiguous. Thoughts?

jwilm commented on src/ansi.rs:
```diff
@@ -742,19 +767,46 @@ impl<'a, H, W> vte::Perform for Performer<'a, H, W>
+            }}
+        }
+
+        let i1: u8 = match intermediates.first() {
```
I think it would make sense to move this into `deg_charset!` since it's not used for anything else. It could then be written like
```rust
macro_rules! deg_charset {
    ($charset:path) => {{
        let index = match intermediates.first() {
            Some(b'(') => CharsetIndex::G0,
            Some(b')') => CharsetIndex::G1,
            Some(b'*') => CharsetIndex::G2,
            Some(b'+') => CharsetIndex::G3,
            _ => unhandled!(),
        };
        self.handler.designate_charset(index, $charset);
    }}
}
```

jwilm commented on src/ansi.rs:
```diff
@@ -742,19 +773,41 @@ impl<'a, H, W> vte::Perform for Performer<'a, H, W>
+            }}
+        }
+
+        macro_rules! conf_charset {
```
Last nit - Can this be expanded to `configure_charset`?

jwilm commented:
This looks great; thanks for all the work here and for following up with reviews! :star2:
